### ISSUE TITLE
Consider using SessionCatalog in SparkTableUtil

### ISSUE THREAD

**Issue by aokolnychyi** [apache/incubator-iceberg#592]
I think it would be better to rely on `SessionCatalog` in `SparkTableUtil` to get a list of partitions. That way, we can support the in-memory catalog as well.

**Comment by aokolnychyi** [comment]
@rdblue is there a particular reason to use the Hive client directly in `SparkTableUtil`?
What about this?

```
spark.sessionState.catalog
  .listPartitions(tableIdentifier)
  .map(p => (p.spec, p.storage.locationUri.map(String.valueOf(_)), p.storage.serde))
  .toDF("partition", "uri", "format")
```

**Comment by rdblue** [comment]
Sounds good to me!

**Comment by aokolnychyi** [comment]
@rdblue, we are also constructing a `DataFrame` in `partitionDF`. What about using a simple case class as we have for manifests and data files? For example, we can have: 

```
case class TablePartition(partition: Map[String, String], uri: String, format: String)
```

**Comment by rdblue** [comment]
Is it possible to use our own HiveCatalog for this instead of using Spark? It would be great to avoid using non-public Spark classes.

**Comment by aokolnychyi** [comment]
Could you elaborate a bit? Not sure I got.

I am proposing to move to `SessionCatalog` in order to get rid of Hive hacks we currently use. In addition, we will get support for the in-memory catalog for free. It also seems reasonable to introduce a case class `TablePartition` like we have `SparkDataFile` as most of the operations in `SparkTableUtil` use typed `Dataset`s. That way, calling `listPartition` will be more readable.

**Comment by rdblue** [comment]
Yeah, those changes make sense to me.

What I was suggesting is maybe to try updating to use the Iceberg HiveCatalog instead of SessionCatalog. Then we wouldn't be relying on an internal API that is not public in Spark (SessionState to get the external catalog).

**Comment by aokolnychyi** [comment]
I hope we won't be using any internal Spark classes to import existing tables at some point. However, that will mean implementing more logic in Iceberg. Also, there are a lot of users who don't use HMS to track partitions. We will also need to cover that use case as well, I believe.

My suggestion is to update `SparkTableUtil` as an intermediate step and create issues to move from internal Spark classes in the future.

**Comment by rdblue** [comment]
That sounds good to me. I just though it might be a good time to take a look if you're changing that area. Thanks for looking into it!

**Comment by aokolnychyi** [comment]
This was resolved in #624
